src/rtc_pl031.rs

```diff
@@ -0,0 +1,617 @@
+        // and that the Data Register RTC count increments over time.
+        let intr_evt = EventFd::new(libc::EFD_NONBLOCK).unwrap();
+        let mut rtc = RTC::new(intr_evt);
+        let mut data = [0; 4];
+
+        // Read the data register.
+        rtc.read(RTCDR, &mut data);
+        let first_read = u32::from_le_bytes(data);
+
+        // Sleep for 1.5 seconds to let the counter tick.
+        let delay = Duration::from_millis(1500);
+        thread::sleep(delay);
+
+        // Read the data register again.
+        rtc.read(RTCDR, &mut data);
+        let second_read = u32::from_le_bytes(data);
+
+        // The second time should be greater than the first
+        assert!(second_read > first_read);
+
+        // Sleep for 1.5 seconds to let the counter tick.
+        let delay = Duration::from_millis(1500);
+        thread::sleep(delay);
+
+        // Writing the data register should have no effect.
+        data = 0u32.to_le_bytes();
+        rtc.write(RTCDR, &data).unwrap();
+
+        // Read the data register again.
+        rtc.read(RTCDR, &mut data);
+        let third_read = u32::from_le_bytes(data);
+
+        // The third time should be greater than the second.
+        assert!(third_read > second_read);
+    }
+
+    #[test]
+    fn test_match_register() {
+        // Test reading and writing to the match register.
+        // TODO: Implement the alarm functionality and confirm an interrupt
+        // is raised when the match register is set.
+        let intr_evt = EventFd::new(libc::EFD_NONBLOCK).unwrap();
+        let mut rtc = RTC::new(intr_evt);
+        let mut data: [u8; 4];
+
+        // Write to and read the value back out of the match register.
+        data = 123u32.to_le_bytes();
+        rtc.write(RTCMR, &data).unwrap();
+        rtc.read(RTCMR, &mut data);
+        assert_eq!(123, u32::from_le_bytes(data));
+    }
+
+    #[test]
+    fn test_load_register() {
+        // Read and write to the load register.
+        let intr_evt = EventFd::new(libc::EFD_NONBLOCK).unwrap();
+        let mut rtc = RTC::new(intr_evt);
+        let mut data = [0; 4];
+
+        // Get the RTC value with a load register of 0 (the initial value).
+        rtc.read(RTCDR, &mut data);
+        let old_val = u32::from_le_bytes(data);
+
+        // Write system time since UNIX_EPOCH in seconds to the load register.
+        let lr = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap()
+            .as_secs();
+        data = (lr as u32).to_le_bytes();
+        rtc.write(RTCLR, &data).unwrap();
+
+        // Read the load register from the load register and verify it
+        // matches the value we just loaded.
+        rtc.read(RTCLR, &mut data);
+        assert_eq!((lr as u32), u32::from_le_bytes(data));
+
+        // Confirm that the new RTC value is greater than the old
+        rtc.read(RTCDR, &mut data);
+        let new_val = u32::from_le_bytes(data);
+        assert!(new_val > old_val);
```

Comment:
I understand that we use a check for monotonically increasing time here since we do not control the time source that the RTC uses during the test. Scheduling delays could mean that system time advances for more than one second while the test is descheduled, thus checking for an exact `RTCDR` value would not be reliable, and it's good that you are checking for a monotonic increase instead.

However, I think the test could be extended; we should make sure that we can also load a *lower* value to `RTCLR` and have it read back successfully from `RTCDR`.

```suggestion
        // Confirm that the new RTC value is greater than the old
        rtc.read(RTCDR, &mut data);
        let new_val = u32::from_le_bytes(data);
        assert!(new_val > old_val);

        // Sleep for 1.5 seconds to let the counter tick.
        let delay = Duration::from_millis(1500);
        thread::sleep(delay);

        // Reset clock back to 0 and confirm it was reset
        let lr = 0;
        data = (lr as u32).to_le_bytes();
        rtc.write(RTCLR, &data).unwrap();
        rtc.read(RTCDR, &mut data);
        assert_eq!((lr as u32), u32::from_le_bytes(data));
```

Reply:
Thinking about it again, the change I suggested suffers from the same problem that the test could be descheduled between the read and the write, and thus trigger a spurious test failure. So it's probably not a good idea to merge my suggested test case without us controlling the RTC time source independent of system time. But it does illustrate the need to reset the counter upon writes to `RTCLR`.

Reply:
If the outcome of the discussion above is that writes to `RTCLR` should reset the offset AND base counter value, then I agree we should find a way to do this (or approximate it...).

Otherwise, I try to perform a check like this in `test_control_register()`.

Reply:
Given that the counter is 1Hz, do you think the scheduler would actually deschedule this for long enough to cause a spurious failure?  I've implemented the test and it's passing, but if it's a real concern we could do two things:
1. Give it three attempts before the test actually fails, assuming the failure mode isn't persistent.
2. Give it a 1 or 2 second buffer.  For example:
```
assert!(u32::from_le_bytes(data) - (lr as u32) <= 1)
```

Reply:
I just pushed an update that is similar to your original recommendation and assumes that we won't be descheduled for more than 1 second.

I can modify to provide either the buffer or the multi-attempt options.

Reply:
Thanks! I think assuming we won't be descheduled for more than 1 second is reasonable. It's a better option than simply omitting this test.

One more suggestion to improve this test specifically: I think we do need a 1.5 second sleep before the write to `RTCLR`. This is to ensure that the counter has ticked at least once, and that we are correctly adjusting the offset if the counter is non-zero. As it is right now, we could erroneously drop the `self.counter = Instant::now();` from the `RTCLR` handler and still pass the test. Adding the sleep protects against this error as well.

Reply:
I get what you're saying now.  I've just pushed an update to the `test_load_register()` test which:
1. Reads the current data register (given an initial load register value of 0).
2. Loads the current time since `UNIX_EPOCH` into the load register.
3. Confirms the data register now reads the time just loaded (assuming we haven't been descheduled for > 1 second!).
4. Confirms the new data register time is greater than the original time from step 1.
4. Loads 0 into the load register.
5. Confirms the data register now reads 0 (again, assuming we haven't been descheduled for > 1 second).